### ISSUE TITLE
Proposal: Split plugin into to segments

### ISSUE THREAD

**Issue by jkroepke** [mahendrapaipuri/grafana-dashboard-reporter-app#41]
Hi,

I would like to leave an proposal that may could help some users, too.

Currently, this plugin requires an chrome instance availible on the grafana container. Typically, the chrome binary isn't part of the grafana container. For this scenario, a custom image is required.

An alternative solution would be to separate the plugin and the renderer process. similar what grafana does with the grafana-image-render.

The benefit would be that no custom image would be required. In Kubernetes, it would be possible do a separate scaling for the rendering process, including auto-scaling based on request.

In short: Instead exec chrome, do an http request to an external web server (separate container) which does the exec.

What did you think about?

**Comment by mahendrapaipuri** [comment]
Thanks for the proposal.

Well, I agree that building a custom image for k8s is not very "convenient". Splitting the plugin would require a major rewrite and will have to manage more things like auth between Grafana and this external service.

Moreover, the plugin delegates most of the heavy work to `grafana-image-renderer` which creates a PNG for each panel in the dashboard. The plugin itself calls chrome at most twice: once to get dashboard model and once to generate report. So, I am not quite sure if scaling is a major issue here. 

Nevertheless it is an interesting idea to pursue in the future if there is a real need for it.

**Comment by jkroepke** [comment]
Would it be possible that the plugin itself ships the chromium binary?

**Comment by mahendrapaipuri** [comment]
That should be possible but we need it test it though. If it works, that would be a great enhancement!

**Comment by jkroepke** [comment]
Hey @mahendrapaipuri 

today, I had another implementation idea

As I know, the developer protocol of chrome supports network.

https://github.com/chromedp/chromedp/blob/ebf842c7bc28db77d0bf4d757f5948d769d0866f/allocate.go#L513-L528

instead using a local ExecAllocation, 

https://github.com/mahendrapaipuri/grafana-dashboard-reporter-app/blob/88d0c230b07ce19f896a6960935ca1a55e7cd959/pkg/plugin/report.go#L323

the `NewRemoteAllocator` can be used to connect against a persistence running chrome instance. This could also safe resource, since it would not be longer required to setup a new chrome instance each time.

**Comment by mahendrapaipuri** [comment]
Hello @jkroepke Thanks a lot for your active involvement. Really appreciate it.

Indeed it is a very good idea. We can use a remote instance "out-of-the-box" for the report generation and I dont really any big blockers here.

But, we use `chromium` to get the "current" dashboard model [here](https://github.com/mahendrapaipuri/grafana-dashboard-reporter-app/blob/88d0c230b07ce19f896a6960935ca1a55e7cd959/pkg/plugin/client.go#L195-L231). As you can see, in this invocation we set _user_ cookie in the chromium to get the dashboard. The rationale is that using user cookie will ensure that the correct RBAC is respected. The blocker here is that we cannot have multiple user cookies in the same instance of chromium. 

One solution is to use [UserDataDir](https://pkg.go.dev/github.com/chromedp/chromedp#UserDataDir) option of `chromedp` and use a different profile for each request so that cookies will be isolated. But again, we need to test it before. 

What do you think?

**Comment by jkroepke** [comment]
> One solution is to use [UserDataDir](https://pkg.go.dev/github.com/chromedp/chromedp#UserDataDir) option of chromedp and use a different profile for each request so that cookies will be isolated. But again, we need to test it before.

I did not expect that this is working. UserDataDir is a start-up option for chrome and I did not expect that this can be changed after chrome is started. 

For an initial POC, I would set the concurrency to 1 (via mutex) that only one PDF can be generated at the same time.

**Comment by mahendrapaipuri** [comment]
> I did not expect that this is working. UserDataDir is a start-up option for chrome and I did not expect that this can be changed after chrome is started.

We dont have to change `UserDataDir`, we have set set a different profile for each user using `chromedp.Flag("profile-directory", username)`. Each user will have their own `profile-directory` and they all will be saved in `UserDataDir`. Does it make sense?

> For an initial POC, I would set the concurrency to 1 (via mutex) that only one PDF can be generated at the same time. and then set the cookie each time.

This can be a solution as well in the worst case.

**Comment by jkroepke** [comment]
Instead playing with cookies/ cookie storage, what about set the cookie header directly?

Like here:

https://github.com/mahendrapaipuri/grafana-dashboard-reporter-app/blob/88d0c230b07ce19f896a6960935ca1a55e7cd959/pkg/plugin/client.go#L210-L211

in theory, cookies are just HTTP headers.

**Comment by mahendrapaipuri** [comment]
> Instead playing with cookies/ cookie storage, what about set the cookie header directly?

Well, when Grafana is hosted on subdomain, browser will send all cookies that are set on the domain to the server. And the fact that operators can configure [cookie name](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#login_cookie_name) will make retrieving correct cookie complicated if we want to set cookie header. That is why we are setting all the cookies in the request.

**Comment by jkroepke** [comment]
The mechanic (just using the cookie header) is already used, I assume, that any mention issues are already covered?

https://github.com/mahendrapaipuri/grafana-dashboard-reporter-app/blob/88d0c230b07ce19f896a6960935ca1a55e7cd959/pkg/plugin/client.go#L98

**Comment by mahendrapaipuri** [comment]
> The mechanic (just using the cookie header) is already used, I assume, that any mention issues are already covered?

I remember testing this strategy as well but didnt work. Probably I didnt look too deep on why it didnt work. 

In any case, if we would like to use one single chrome instance, we **must** use different profile for each user to avoid any accidental leak of sensitive info in either case setting cookie header or cookies themselves. Don't you agree?

**Comment by jkroepke** [comment]
I do a local test:

<details>
<summary>Details</summary>


```go
// Command headers is a chromedp example demonstrating how to add extra HTTP
// headers to browser requests.
package main

import (
	"context"
	"fmt"
	"log"
	"os"

	"github.com/chromedp/cdproto/network"
	"github.com/chromedp/cdproto/storage"
	"github.com/chromedp/chromedp"
)

func main() {
	dir, err := os.MkdirTemp("", "chromedp-example")
	if err != nil {
		log.Fatal(err)
	}
	defer os.RemoveAll(dir)

	opts := append(chromedp.DefaultExecAllocatorOptions[:],
		chromedp.DisableGPU,
		chromedp.UserDataDir(dir),
	)

	ctx, cancel := chromedp.NewExecAllocator(context.Background(), opts...)
	defer cancel()

	// new browser, first tab
	ctx1, cancel1 := chromedp.NewContext(ctx, chromedp.WithDebugf(log.Printf))
	defer cancel1()

	var (
		res1 string
		res2 string
		res3 string
	)

	// ensure the first tab is created
	if err := chromedp.Run(ctx1, setheaders(
		"https://httpbin.org/cookies",
		map[string]interface{}{
			"Cookie": "name=1value; name2=1value2; name3=1value3",
		},
		&res1,
	)); err != nil {
		log.Fatal(err)
	}

	// same browser, second tab
	ctx2, _ := chromedp.NewContext(ctx1)

	// ensure the second tab is created
	if err := chromedp.Run(ctx2, setheaders(
		"https://httpbin.org/cookies",
		map[string]interface{}{
			"Cookie": "name=2value; name2=2value2; name3=2value3",
		},
		&res2,
	)); err != nil {
		log.Fatal(err)
	}

	// same browser, thrid tab
	ctx3, _ := chromedp.NewContext(ctx1)

	// ensure the second tab is created
	if err := chromedp.Run(ctx3, setheaders(
		"https://httpbin.org/cookies",
		map[string]interface{}{},
		&res3,
	)); err != nil {
		log.Fatal(err)
	}

	fmt.Printf("%+v\n", res1)
	fmt.Printf("%+v\n", res2)
	fmt.Printf("%+v\n", res3)
}

// setheaders returns a task list that sets the passed headers.
func setheaders(host string, headers map[string]interface{}, res *string) chromedp.Tasks {
	return chromedp.Tasks{
		network.Enable(),
		network.SetExtraHTTPHeaders(network.Headers(headers)),
		chromedp.Navigate(host),
		chromedp.Text(`body > pre`, res),
		chromedp.ActionFunc(func(ctx context.Context) error {
			cookies, err := storage.GetCookies().Do(ctx)
			if err != nil {
				return err
			}

			log.Printf("chrome cookie: %+v", cookies)

			return nil
		}),
		network.ClearBrowserCookies(),
		chromedp.Stop(),
	}
}
```

</details>

In that test, I'm do multiple REST calls within one browser. I set cookie via `network.SetExtraHTTPHeaders`. 

I could observe that the header is only used inside a browser tab. The Tab3 did not have have access to the cookies from Tab1.

Cookies are not persistent stored inside chrome. `storage.GetCookies()` reports 0 cookies after that.

Based on that given test, I do not have the feeling that dedicated browser instance with separate profiles are not necessary. Browser contexts (aka tabs) are sufficient to separat requests from each other, since no cookies are persistent. However we can also enforce ClearCookies action after all.

**Comment by mahendrapaipuri** [comment]
Thanks a lot for gist @jkroepke 

Well, we can give a another go using cookie header instead setting cookies. If we going to use the chrome instance in a separate container (in k8s context), we need to look on how to secure the access to this instance.

**Comment by jkroepke** [comment]
remote debugging protocol doesn't support authentication and TLS.

the most secure way would running the dedicated chrome container inside them Kubernetes pod. Grafana and Chrome still share the same network namespace and chrome would be reachable via localhost or unix socket.

**Comment by mahendrapaipuri** [comment]
I suppose Grafana Image Renderer when running in remote mode uses sidecar architecture as well. There was a severe [CVE](https://github.com/grafana/grafana-image-renderer/security/advisories/GHSA-2cfh-233g-m4c5) reported last year and Grafana team proposed a patch by using a shared secret between Grafana and remotely running Grafana image renderer. This is what bothers me and I am pretty sure attack vectors like this will be possible when running chrome instance in a separate side car container with any auth between them.

**Comment by jkroepke** [comment]
The Grafana image renderer is running as dedicated pod (not as sidecar) and is exposing all network capabilities.

My idea was to run the chrome for grafana-dashboard-reporter-app as sidecar which does not require to expose the network capabilities. The chrome can listen on localhost only. 

And unlike grafana-image-renderer, grafana-dashboard-reporter-app does not permanently store any credentials. If a remote attacker gets access to an chrome instance, he just finds an empty chrome instance.

**Comment by mahendrapaipuri** [comment]
> The Grafana image renderer is running as dedicated pod (not as sidecar) and is exposing all network capabilities.

Ah, I wasnt aware of that. Thanks for detail!

> The chrome can listen on localhost only.

That should be reassuring. That should prevent a lot of attack vectors already.

>  If a remote attacker gets access to an chrome instance, he just finds an empty chrome instance.

I am not quite sure. As it is a remote debugging protocol, I guess (not entirely sure), the attacker can introspect the current open tabs and potentially retrieve secrets. But as you said, if chrome instance will be on localhost with respect to Grafana main container, the chances of attacker getting hold of chrome instance is very low.

**Comment by jkroepke** [comment]
One more thing: In Kubernetes it's possible to setup an volume to share file between containers.

Render a local html file

https://github.com/mahendrapaipuri/grafana-dashboard-reporter-app/blob/88d0c230b07ce19f896a6960935ca1a55e7cd959/pkg/plugin/report.go#L336

would still work and is not a blocker here.

**Comment by mahendrapaipuri** [comment]
If we are moving towards remote chrome instance approach, we can do all the processing in-memory by converting PNG to base64 encoded strings. That will avoid all sorts of file system access which can pose problems in k8s context.

**Comment by jkroepke** [comment]
temporary file system access does create any issues. End users just have to create an empyDir Volume and kubernetes will take care off all correct permissions.

The step you mention is an general improvement, and not strictly necessary for the first milestone.
